### ISSUE TITLE
how to apply a color gradient to the underlining of a text

### ISSUE THREAD

**Issue by asturur** [fabricjs/fabric.js#9726]
### Discussed in https://github.com/fabricjs/fabric.js/discussions/9665

<div type='discussions-op-text'>

<sup>Originally posted by **Pablituuu** February 12, 2024</sup>
You can apply a gradient color to text without problems, but when applying underline or linethrough to the text, its color is not applied to these options. Only if the text color is solid can it be applied. Is there a solution for this?

I leave the link of the reproduction:
https://jsfiddle.net/Pablituuu/8h3j7L26/39/</div>

**Comment by mspanish** [comment]
Seems like this would be related, so I'll add it here. It would be great to be able to set the color for underlining. I'm building a note taking app, and underlining something in red for example would be a handy feature.

**Comment by asturur** [comment]
Yes would be nice at least for underline and strike. Not sure why we have overline tho.
In order for this to work a series of decision needs to be taken, like how do i target the underine at a specific place in the text style, underlineFIll? and why underline has no fill and stroke? and then there is risk of a lot of extra features.

**Comment by asturur** [comment]
Is also not hard to build, someone could give a stab to it if they are not afraid or bothered by long pr review process

**Comment by mspanish** [comment]
Overline is actually kind of cool as it makes a block of text stand out in your note taking (used with underline is what has kind of a cool effect). Is there any way to hack this to add it quickly? I see the fn that applies it is   _renderTextDecoration  - but since 6+ I'm not really sure how I could override that, the new syntax kind of throws me.

**Comment by MicleMing** [comment]
It seems I have a related need as well — I want to achieve a highlight-like effect, which requires customization. How can I achieve it?

<img width="151" height="68" alt="Image" src="https://github.com/user-attachments/assets/ec941452-0e24-4722-959d-937438d33833" />

**Comment by asturur** [comment]
This gradient thing should be done for fabric 7 since is a breaking change ( i think )
For the highlight style effect i think all you can do is to override the `_render` method of text:

```diff
  /**
   * @private
   * @param {CanvasRenderingContext2D} ctx Context to render on
   */
  _render(ctx: CanvasRenderingContext2D) {
    const path = this.path;
    path && !path.isNotVisible() && path._render(ctx);
    this._setTextStyles(ctx);
    this._renderTextLinesBackground(ctx);
-   this._renderTextDecoration(ctx, 'underline');
+   your magic code
    this._renderText(ctx);
    this._renderTextDecoration(ctx, 'overline');
    this._renderTextDecoration(ctx, 'linethrough');
  }
```

and taking inspiration from how _renderTextDecoration measures what and where to draw, write your own function.
Not sure how that would look like to be honest.
